ecosystem/gstreamer_plugin/gst_mtl_st20p_tx.c

```diff
@@ -543,3 +571,101 @@ GST_PLUGIN_DEFINE(GST_VERSION_MAJOR, GST_VERSION_MINOR, mtl_st20p_tx,
                   GST_PACKAGE_ORIGIN)
+
+static int gst_mtl_st20p_tx_frame_done(void* priv, struct st_frame* frame) {
+  if (frame == NULL || frame->opaque == NULL) {
+    return 0;
+  }
+
+  st20_tx_external_data_child_t* child = frame->opaque;
+  st20_tx_external_data_parent_t* parent = child->parent;
+
+  gst_memory_unmap(child->gst_buffer_memory, &child->map_info);
+
+  frame->opaque = NULL;
+  free(child);
+
+  pthread_mutex_lock(&parent->parent_mutex);
+  parent->child_count--;
+  if (parent->child_count > 0) {
+    pthread_mutex_unlock(&parent->parent_mutex);
+    return 0;
+  }
+
+  pthread_mutex_unlock(&parent->parent_mutex);
+  gst_buffer_unref(parent->buf);
+  pthread_mutex_destroy(&parent->parent_mutex);
+  free(parent);
+
+  return 0;
+}
+
+static GstFlowReturn gst_mtl_st20p_tx_zero_copy(Gst_Mtl_St20p_Tx* sink, GstBuffer* buf) {
+  GstVideoMeta* video_meta = gst_buffer_get_video_meta(buf);
+  gint buffer_n = gst_buffer_n_memory(buf);
+  if (!video_meta) {
+    g_print("Failed to get video meta from buffer\n");
+    return GST_FLOW_ERROR;
+  }
+
+  st20_tx_external_data_parent_t* parent = malloc(sizeof(st20_tx_external_data_parent_t));
+  if (!parent) {
+    GST_ERROR("Failed to allocate memory for parent structure");
+    return GST_FLOW_ERROR;
+  }
+  parent->buf = buf;
+  parent->child_count = buffer_n;
+  pthread_mutex_init(&parent->parent_mutex, NULL);
+
+  for (int i = 0; i < buffer_n; i++) {
+    st20_tx_external_data_child_t* child = malloc(sizeof(st20_tx_external_data_child_t));
+    if (!child) {
+      GST_ERROR("Failed to allocate memory for child structure");
+      free(parent);
+    }
+    child->parent = parent;
+    child->gst_buffer_memory = gst_buffer_peek_memory(buf, i);
+
+    if (!gst_memory_map(child->gst_buffer_memory, &child->map_info, GST_MAP_READ)) {
+      GST_ERROR("Failed to map memory");
+      free(child);
+      free(parent);
+      return GST_FLOW_ERROR;
+    }
+
+    if (child->map_info.size < sink->frame_size) {
+      GST_ERROR("Buffer size %lu is smaller than frame size %d", child->map_info.size,
+                sink->frame_size);
+      gst_memory_unmap(child->gst_buffer_memory, &child->map_info);
+      free(child);
+      free(parent);
+      return GST_FLOW_ERROR;
+    }
+
+    struct st_frame* frame = st20p_tx_get_frame(sink->tx_handle);
+    if (!frame) {
+      GST_ERROR("Failed to get frame");
+      return GST_FLOW_ERROR;
+    }
+
+    // By default, timestamping is handled by MTL.
+    if (sink->use_pts_for_pacing) {
+      frame->timestamp = GST_BUFFER_PTS(buf) += sink->pts_for_pacing_offset;
+      frame->tfmt = ST10_TIMESTAMP_FMT_TAI;
+    }
+
+    struct st_ext_frame ext_frame;
+    for (int i = 0; i < video_meta->n_planes; i++) {
+      ext_frame.addr[i] = child->map_info.data + video_meta->offset[i];
+      ext_frame.linesize[i] = video_meta->stride[i];
+      ext_frame.iova[i] = 0;
+    }
+    ext_frame.size = child->map_info.size;
+    ext_frame.opaque = child;
```

Comment:
Does this work? Here you set `opaque` in `struct st_ext_frame` but in `gst_mtl_st20p_tx_frame_done` you read it from `struct st_frame` is it moved to there?

Reply:
Yes, it works, MTL copy opaque from ext_frame to st_frame